MBS-10866 / MBS-11409: Call TO_JSON explicitly in all cases

mwiencek opened the pull request:
Fixes MBS-10866 & MBS-11409.

Prior to this patch, there was a longstanding issue whereby the encoding of the `$linked_entities` hash would cause modifications to it while it was being encoded. This would happen because in many cases, we've relied on convert_blessed to have the encoder call `TO_JSON` for us -- and any TO_JSON method could of course call add_linked_entity again.

As the error message in MBS-11409 hints at, this caused undefined behavior. There was no guarantee that an entity added to `$linked_entities` mid-encoding would appear in the final encoded output. This caused mysterious bugs like MBS-10866.

The use of `convert_blessed` was convenient because it meant we could pass in blessed entities without needing to worry about converting them to plain hash refs. Unfortunately, there's no safe way to guarantee that `$linked_entities` is fully populated -- and won't be modified during encoding -- without explicitly calling TO_JSON everywhere beforehand.

This is an invasive change, because it affects so many different places across the site:

  * TO_JSON methods (entity layer)
  * component_props (controller layer)
  * React.embed props (TT)
  * build_display_data (edits)
  * inflate_rows (reports)
  * ???

I tested this by browsing as many pages on the site as I could: clicking on all tabs of every entity type, loading at least one edit of every type, performing various edit searches, etc. Further testing on test.mb and beta.mb is recommended.

reosarevok commented on lib/MusicBrainz/Server/Report/PlaceReport.pm:
```diff
@@ -3,6 +3,13 @@ use Moose::Role;
 
 with 'MusicBrainz::Server::Report::QueryReport';
 
+sub _load_extra_place_info {
+    my ($self, @places) = @_;
+
+    $self->c->model('Area')->load(@places);
+    $self->c->model('Area')->load_containment(map { $_->area } @places);
+}
+
```
Is this for clarity, or is it a direct requirement of the JSON changes? I like it, anyway, it makes it a lot more clear, I just wasn't sure if it's required.

mwiencek replied:
It's a requirement because IIRC we had some reports overriding `inflate_rows` from the parent and loading more info on the output -- which is now impossible because `inflate_rows` returns plain hash refs, not blessed entities.

reosarevok commented on root/edit/details/AddArea.js:
```diff
@@ -27,7 +27,7 @@ type AddAreaEditT = {
     +name: string,
     +sort_name: string | null,
-    +type: AreaTypeT,
+    +type: AreaTypeT | null,
```
I guess this isn't new, but an old typing error? :)

mwiencek replied:
Yeah. I'll go back and try to split out some of the changes you've brought up into separate commits.

reosarevok reviewed (commented):
Some (minor) comments after looking through the code. Will test through the site and comment more if I find issues.

reosarevok commented on lib/MusicBrainz/Server/Report/CDTOCReport.pm:
```diff
@@ -22,8 +22,8 @@ around inflate_rows => sub {
             %$_,
-            cdtoc => $cdtocs->{ $_->{cdtoc_id} },
-            release => $releases->{ $_->{release_id} }
+            cdtoc => $cdtocs->{ $_->{cdtoc_id} }->TO_JSON,
+            release => $releases->{ $_->{release_id} }->TO_JSON,
```
More generally, is there a reason to use TO_JSON directly above object_to_json?

mwiencek replied:
No reason, I wasn't thinking of that case and assumed they would always exist. Will fix.

mwiencek replied:
Changed all reports to use `to_json_object`.

reosarevok commented on lib/MusicBrainz/Server/Report/EditorReport.pm:
```diff
@@ -16,7 +16,7 @@ around inflate_rows => sub {
         map +{
             %$_,
-            editor => $editors->{ $_->{id} }
+            editor => $editors->{ $_->{id} }->TO_JSON,
```
This is not new from this patch, but can you fix MBS-11423 while you're changing this? It needs unsanitized data (emails).

mwiencek commented on lib/MusicBrainz/Server/Entity/Util/JSON.pm:
```diff
@@ -55,11 +58,11 @@ sub encode_with_linked_entities {
-    my $linked_entities = $json_encoder->encode($linked_entities);
-    $encoded =~ s/}$/,"linked_entities":$linked_entities}/;
+    if (defined $linked_entities) {
+        my $encode_linked_entities = $json_encoder->encode($linked_entities);
```
Yes, that's what I meant. :) I'll fix it.

mwiencek commented on lib/MusicBrainz/Server/Edit.pm:
```diff
@@ -341,7 +342,7 @@ sub TO_JSON {
         is_open => boolean_to_json($self->is_open),
         $can_preview ? (preview => boolean_to_json($self->preview)) : (),
         status => $self->status + 0,
-        votes => [map { $_->TO_JSON } $self->all_votes],
+        votes => to_json_array($self->votes),
```
`all_votes` returns an array (i.e. list), `votes` returns an array reference. `to_json_array` works on references.

The reason I made `to_json_array` take an array ref and not an array is that it can handle cases where it receives a scalar value like `undef` (and just return that back). If it took an array, it would have no way to distinguish `undef` from a list of one element.

mwiencek commented on lib/MusicBrainz/Server/Edit/Relationship/Edit.pm:
```diff
@@ -240,6 +241,10 @@ sub build_display_data {
     my $old = $self->data->{old};
     my $new = $self->data->{new};
 
+    $self->c->model('LinkAttributeType')->load(map {
+        $_->all_attributes
+    } values %{ $loaded->{LinkType} });
+
```
I was confused by this too. I found an edit that failed to load without it, so it might have been an unrelated, existing bug. The issue was that linkPhrase.js needed access to all the attributes associated with the link type, IIRC. I'll have to find what edit that was.

mwiencek replied:
Figured this out. The line in `Entity::LinkType` where we do `$self->link_entity('link_attribute_type', $_->type_id, $_->type);` was storing `undef` for those link attribute types, because they weren't loaded on the link type. And my commit titled "Further avoid infinite recursion in add_linked_entity" broke things because I changed the `!defined` check to `!exists`, so the defined ones from the relationship weren't added due to the `undef` values stored from the link type earlier.

Fixed it by changing that commit to check if `$entity` is defined before storing it. Reverted this.

mwiencek commented on lib/MusicBrainz/Server/Entity/Role/Linkable.pm:
```diff
@@ -75,13 +75,20 @@ sub relationships_by_link_type_names
 
-    if ($self->has_loaded_relationships) {
-        $json->{relationships} = [map { $_->TO_JSON } $self->all_relationships];
+    if ($_serialize_relationships && $self->has_loaded_relationships) {
+        $json->{relationships} = [map {
+            $_serialize_relationships = 0;
+            my $result = $_->TO_JSON;
+            $_serialize_relationships = 1;
+            $result;
+        } $self->all_relationships];
```
You can have cyclic references via relationships. Like Artist1.relationships @> Work1 and Work1.relationships @> Artist1. Calling TO_JSON on Artist1 would serialize the Work1 relationship, which would serialize Work1, then Work1.relationships, then Artist1 again, leading to infinite recursion.

It wasn't a problem before, because the serialization was lazy -- by the time we got around to serializing Work1, Artist1 would have already been added to `$linked_entities`, so it would be skipped on the occurrence.

The commit titled "Further avoid infinite recursion in add_linked_entity" (I won't link to it directly because I'll likely break the commit ID later) doesn't fix this because we don't output relationship targets via `$linked_entities`, we just output them directly.

mwiencek replied:
After looking into this again, I can't find where I triggered this kind of infinite recursion. I did somewhere, but...

I'll keep this check just in case, but take the explanation above with a grain of salt. It's safer to keep around until we move relationship targets to `$linked_entities`.

mwiencek commented on lib/MusicBrainz/Server/Entity/Work.pm:
```diff
@@ -95,13 +96,13 @@ around TO_JSON => sub {
-        iswcs => [map { $_->TO_JSON } $self->all_iswcs],
-        artists => [map { $_->TO_JSON } $self->all_artists],
+        attributes => to_json_array([$self->sorted_attributes]),
+        languages => to_json_array($self->languages),
+        iswcs => to_json_array($self->iswcs),
+        artists => to_json_array($self->artists),
```
See https://github.com/metabrainz/musicbrainz-server/pull/1958#discussion_r588446757

mwiencek commented on lib/MusicBrainz/Server/Edit/Instrument/DeleteAlias.pm:
```diff
@@ -43,8 +44,10 @@ around 'build_display_data' => sub {
-    $data->{instrument} = $loaded->{Instrument}->{ $self->instrument_id }
-        || Instrument->new(name => $self->data->{entity}{name});
+    $data->{instrument} = to_json_object(
+        $loaded->{Instrument}->{ $self->instrument_id } ||
```
I don't know if it's worth it, but I'll change the ones in `build_display_data` at least in a separate commit (won't touch the others 'cause it'll make the PR much larger than it already is).

reosarevok commented on lib/MusicBrainz/Server/Report/WorkReport.pm:
```diff
@@ -13,6 +15,8 @@ around inflate_rows => sub {
         map { $_->{work_id} } @$items
     );
 
+    $self->_load_extra_work_info(values %$works);
```
Is this useful to have for a `{}` function? Is it just for future usage?

mwiencek replied:
It's used currently by `MusicBrainz::Server::Report::ISWCsWithManyWorks`.

reosarevok replied:
Oh, sure, it's for around or whatnot. Sorry, my bad.

reosarevok reviewed (changes requested):
http://localhost:5000/artist/9ddd7abc-9e1b-471d-8031-583bc6bc8be9 (on `pink`) has the following annotation message:
`This is an old revision of this annotation, as edited by [missing editor] on 2005-12-20 19:40 GMT+2. View current revision.`
Beta has 
`Annotation last modified by Shepard on 2005-12-20 19:40 GMT+2. Edit annotation"

entity.latest_annotation seems to be missing once the hydration kicks in, causing this.

mwiencek commented:
Thanks, fixed. The issue was that in `Controller::Role::Annotation`, we load the editor in `after 'show'`, which is now *after* we convert the annotation to JSON. I changed it to `before 'show'` and it works again. I search for other places we load stuff in `after 'show'` and found a similar issue with Wikipedia extracts: we were outputting `null` for the cached version, though this bug wasn't as obvious since we'd just load it asynchronously in the browser in that case.

reosarevok reviewed (approved):
This is probably fine to put on beta for proper full-site testing once conflicts are fixed - I didn't see any other issues while navigating around but it might be we will only find some hidden ones when users use it normally.

yvanzo reviewed (approved):
Looks great to me! I proofread most of it and tested as much as I could, but we really need beta  for this one more than for any other patchset as @reosarevok said..
